docs/modules/password_auth_provider_callbacks.md

```diff
@@ -105,6 +105,64 @@ device ID), and the (now deactivated) access token.
 
 If multiple modules implement this callback, Synapse runs them all in order.
 
+### `get_username_for_registration`
+
+_First introduced in Synapse v1.52.0_
+
+```python
+async def get_username_for_registration(
+    uia_results: Dict[str, Any],
+    params: Dict[str, Any],
+) -> Optional[str]
+```
+
+Called when registering a new user, allowing the module to define the username to give the user.
+The username is the local part of a user's full Matrix ID, i.e. it's `alice` in `@alice:example.com`.
+
+This callback is called once [User-Interactive Authentication](https://spec.matrix.org/latest/client-server-api/#user-interactive-authentication-api)
+has been completed by the user. It is passed two dictionaries, which include the information that the user has provided
+during the registration process.
+
+The first dictionary contains the results of the [User-Interactive Authentication](https://spec.matrix.org/latest/client-server-api/#user-interactive-authentication-api)
+flow followed by the user. Its keys are the identifiers of every step involved in the flow,
+associated with either a boolean value indicating whether the step was correctly completed,
+or additional information (e.g. email address, phone number...). Here's an example featuring
+all currently supported keys:
+
+```python
+{
+    "m.login.dummy": True,  # Dummy authentication
+    "m.login.terms": True,  # User has accepted the terms of service for the homeserver
+    "m.login.recaptcha": True,  # User has completed the recaptcha challenge
+    "m.login.email.identity": {  # User has provided and verified an email address
+        "medium": "email",
+        "address": "alice@example.com",
+        "validated_at": 1642701357084,
+    },
+    "m.login.msisdn": {  # User has provided and verified a phone number
+        "medium": "msisdn",
+        "address": "33123456789",
+        "validated_at": 1642701357084,
+    },
+    "org.matrix.msc3231.login.registration_token": "sometoken",  # User has registered through the flow described in MSC3231
+}
```

Comment:
Is a link to https://spec.matrix.org/v1.1/client-server-api/#authentication-types useful here?

I couldn't see `m.login.terms` in that link, and I couldn't see `m.login.sso` in this example. I don't know if that's actually a concern though.

Reply:
SSO isn't included because it actually happens in a separate code path (and I'm not sure having modules interact with SSO this way is a good idea) - I'll add a precision about this in the doc. `m.login.terms` is https://github.com/matrix-org/matrix-doc/pull/1692 which is (*sigh*) still not merged.

I'll add a link but will specify that this includes _most_ auth types, not all of them.